### ISSUE TITLE
[HELP/Guide] - Some Trivial Questions

### ISSUE THREAD

**Issue by pknmax** [graphdeco-inria/gaussian-splatting#505]
If I have following patch (consider it as planar object) with view dependent color,
![image](https://github.com/graphdeco-inria/gaussian-splatting/assets/152391989/b399953a-ed0d-48eb-9e37-ba275cc93e59)

and I captured this patch from different directions, trained with 3DGS, & got the final .ply file.

1. How many gaussians (approx) will it need to learn this patch with view dependent colors?
2. How I can get those view dependent colors from final .ply file in RGB format given viewing direction?

**Comment by Ph03n1xdust** [comment]
Depends on how accurately do you want to capture the edges of it. If it is really a "plane" then 1 gaussian should be enough. With huge X,Y scale and small Z scale.

You can check the how the `_features_dc` and `_features _rest` are extracted from the ply files with the `load_ply` function  in `gaussian_model.py`. And then how to convert them to view dependent RGB in `sh_utils.py`.

**Comment by pknmax** [comment]
@Ph03n1xdust 
thanks for quick reply.
I was checking your suggestion using __init__.py of the gaussian_renderer,
but how I can get viewing direction (viewpoint_camera) from .ply file?
```python
    shs = None
    colors_precomp = None
    if override_color is None:
        if pipe.convert_SHs_python:
            shs_view = pc.get_features.transpose(1, 2).view(-1, 3, (pc.max_sh_degree+1)**2)
            dir_pp = (pc.get_xyz - viewpoint_camera.camera_center.repeat(pc.get_features.shape[0], 1))
            dir_pp_normalized = dir_pp/dir_pp.norm(dim=1, keepdim=True)
            sh2rgb = eval_sh(pc.active_sh_degree, shs_view, dir_pp_normalized)
            colors_precomp = torch.clamp_min(sh2rgb + 0.5, 0.0)
        else:
            shs = pc.get_features
    else:
        colors_precomp = override_color
```

**Comment by Ph03n1xdust** [comment]
The ply file does not contain the viewing direction. It depends on where you are actually viewing the gaussians from :) 

The `viewpoint_camera` is the camera object you are using to render the image. See a bit more discussion about it in #506

**Comment by pknmax** [comment]
@Ph03n1xdust  Thank you.

Now suppose I have a .ply files for individual color patches like below (one .ply file for one color), which changes color according to viewing angle, 
![image](https://github.com/graphdeco-inria/gaussian-splatting/assets/152391989/3d86f654-3d11-4a67-8f24-9a1fcf7955ea)

How can I create a .ply for a planar object/image(3D) from above learned .ply files of individual color patches and location of these learned colors in my 3D planar object/image?

**Comment by Ph03n1xdust** [comment]
You can read them with any ply file reading package you want, and simply put all of them in the same ply file and save that, with changing the x,y,z coordinates you can move them w.r.t. each other. If you actually want to rotate them that is more tricky, but there were some questions regarding this in one of the issues (I can't find it right now).

Another approach is that you just load them with the GaussianModel-s from the 3DGS code, and make a new model where you set the _xyz, _opacity, ..... concatenating all the stuff you want to put together and the GaussianModel also have a save_ply method which you can use to save it.

**Comment by pknmax** [comment]
To make earlier question clear:
Suppose I want to generate planar 3D image of following image using learned .ply file above learned color patches, considering I know the location of these learned colors in my planar 3D image, how I can do it and is it possible?
![image](https://github.com/graphdeco-inria/gaussian-splatting/assets/152391989/e4abd882-a55d-406c-a2bf-3e47ec406f9c)

following difficulties I am facing:
1. I can not simply capture images of patches as colmap fails to compute cameras due to lack texture or mapping points in the capture images. So I am using chessboard under the color patches while capturing.
2. Also 3DGS adds gaussians at back and sometimes at front of color patch to generate view dependent colors. (So I guess one gaussian is not enough to capture view dependent color patch.)
3. Due above to reasons I need to crop the unnecessary gaussians from .ply but it changes appearance of gaussians which represents color patch.
4. Other solution is, query the colors of learned color patches from .ply file for single viewing direction and pace it in the target 2D image color map and simulate.
5. Is there any better solution?

**Comment by pknmax** [comment]
> I am a bit at loss on what you want to achieve exactly with this. Do you want to make a 2D object (which can be placed in a 3D environment) and when you "walk" in front of it it should change from one picture to another? Like a picture on a wall, but it looks different depending on where you look at it from?

Initially, I want to simulate a 3D planar image generated using view dependent colors (which are learned from those view dependent color patches and these view dependent colors are shiny/glossy/specular) and I know the location of these view dependent colors in the target 3D planar image. 

So when I view this generated 3D planar image in SIBR viewer it should change the color when I rotate it/change the viewing direction.

**Comment by Ph03n1xdust** [comment]
Does your checkerboard COLMAP model correctly place the points inside the square you want to create?

If yes (I assume it does, and that your patch lives in the XY plane), then you can simply train your GS, but force it to stay in plane:
- set the z scaling to a very small number every iteration
- force the rotation to be an identity

This should result in a more or less good description of the patch you wanted in the first place. (If the patch is not properly reconstructed you can use a denser checkerboard/place more points by hand inside the rectangle before training)

Once you have the correct patch you can build up whatever you want from them, by simply placing them. (There were also some issues about building the ply files by hand, you can find those)

**Comment by pknmax** [comment]
> If you have already trained a model from that colmap dataset in the same folder the GS generates a ply file, and if the ply file is there the the code will use that ply file and not re-read the points3D.bin/txt so every time you change this file, you should also delete the ply from next to it.

I can see the input.ply file in output folder generated after training is what I intended to initialize the gaussians with but when I see initial point cloud in SIBR viewer it shows the original point cloud generated by colmap.

**Comment by pknmax** [comment]
> The cfg_args contain the path of the original training data, maybe it loads from there. Other than that, I have no clue, sorry.

I guess you are correct.

Another question:
Can you please tell me or point out to similar issue on how I can build ply file by hand (I could not find any related issue you mentioned [earlier](https://github.com/graphdeco-inria/gaussian-splatting/issues/505#issuecomment-1843411999)):
1. Consider I want replace 0 and 1 in binary image with two different gaussians (and associated colors) from other ply files to generate 3D planar image.
2. How I can achieve same for rgb image?
